core-bundle/src/Routing/ResponseContext/ResponseContext.php

```diff
@@ -12,31 +12,46 @@
 
 namespace Contao\CoreBundle\Routing\ResponseContext;
 
-use Symfony\Component\HttpFoundation\Response;
-use Symfony\Component\HttpFoundation\ResponseHeaderBag;
-
-class ResponseContext implements ResponseContextInterface
+final class ResponseContext
 {
+    public const REQUEST_ATTRIBUTE_NAME = '_contao_response_context';
+
+    /**
+     * @var array
+     */
+    private $services = [];
+
     /**
      * @var PartialResponseHeaderBag|null
      */
     private $headerBag;
 
-    public function getHeaderBag(): ResponseHeaderBag
+    public function add(object $service): self
     {
-        if (null === $this->headerBag) {
-            $this->headerBag = new PartialResponseHeaderBag();
-        }
+        $this->services[\get_class($service)] = $service;
 
-        return $this->headerBag;
+        return $this;
+    }
+
+    public function has(string $service): bool
+    {
+        return null !== $this->get($service);
     }
 
-    public function finalize(Response $response): ResponseContextInterface
+    /**
+     * @return mixed|null
+     */
+    public function get(string $service)
     {
-        foreach ($this->getHeaderBag()->all() as $name => $values) {
-            $response->headers->set($name, $values, false); // Do not replace but add
+        return $this->services[$service] ?? null;
```

Comment:
I'd prefer if we would throw an exception when a service does not exist. Then the result would always be an object and - together with the template annotations above - you could do something like that:

```php
$context->get(HeadBag::class)->setTitle('foo'); // <-- autocompletion here
```

IMO this would closely resemble how `Container#get()` works by default in Sf.